File: backend/app/schemas/parental_controls.py
"""Pydantic schemas for parental controls endpoints."""

import uuid
from datetime import datetime
from typing import ClassVar
from zoneinfo import available_timezones

from pydantic import BaseModel, Field, field_validator


# -- PIN schemas --

class PinCreate(BaseModel):
    new_pin: str = Field(min_length=4, max_length=4, pattern=r"^\d{4}$")
    current_pin: str | None = Field(None, min_length=4, max_length=4, pattern=r"^\d{4}$")


class PinVerify(BaseModel):
    pin: str = Field(min_length=4, max_length=4, pattern=r"^\d{4}$")


class PinReset(BaseModel):
    password: str = Field(min_length=1)
    new_pin: str = Field(min_length=4, max_length=4, pattern=r"^\d{4}$")


class PinResponse(BaseModel):
    detail: str


class PinError(BaseModel):
    detail: str
    # M-09: remaining_attempts removed — don't leak attempt count to clients
    locked_until: datetime | None = None


class PinVerifyResponse(BaseModel):
    verified: bool
    pin_token: str | None = None


class PinStatusResponse(BaseModel):
    has_pin: bool
    locked_until: datetime | None = None


# -- Viewing Time Config schemas --

class ViewingTimeConfigResponse(BaseModel):
    profile_id: uuid.UUID
    weekday_limit_minutes: int | None
    weekend_limit_minutes: int | None
    reset_hour: int
    educational_exempt: bool
    timezone: str
    updated_at: datetime

    model_config = {"from_attributes": True}


class ViewingTimeConfigUpdate(BaseModel):
    weekday_limit_minutes: int | None = Field(None, ge=15, le=480)
    weekend_limit_minutes: int | None = Field(None, ge=15, le=480)
    reset_hour: int | None = Field(None, ge=0, le=23)
    educational_exempt: bool | None = None
    timezone: str | None = Field(None, max_length=50)
    pin_token: str | None = Field(None,
                                  description="Required when changing sensitive fields (reset_hour, educational_exempt)")

    # M-03: Fields that require PIN verification to change
    SENSITIVE_FIELDS: ClassVar[set[str]] = {"reset_hour", "educational_exempt"}

    @field_validator("timezone")
    @classmethod
    def validate_timezone(cls, v: str | None) -> str | None:
        if v is not None and v not in available_timezones():
            raise ValueError(f"Invalid timezone: {v}")
        return v


# -- Grant Extra Time schemas --

class GrantExtraTimeRequest(BaseModel):
    minutes: int | None = Field(None, ge=15, le=120,
                                description="15, 30, or 60 minutes. null = Unlimited for today.")
    pin: str | None = Field(None, min_length=4, max_length=4, pattern=r"^\d{4}$",
                            description="Inline PIN verification for on-device grant")
    pin_token: str | None = Field(None,
                                  description="Short-lived token from /pin/verify (alternative to pin)")


class GrantExtraTimeResponse(BaseModel):
    granted_minutes: int | None
    remaining_minutes: float | None
    is_unlimited_override: bool


# -- Viewing History schemas --

class ViewingHistorySession(BaseModel):
    session_id: uuid.UUID
    title_id: uuid.UUID
    title_name: str
    device_type: str | None
    is_educational: bool
    started_at: datetime
    ended_at: datetime | None
    duration_minutes: float


class ViewingHistoryDay(BaseModel):
    date: str
    total_minutes: float
    educational_minutes: float
    counted_minutes: float
    sessions: list[ViewingHistorySession]


class ViewingHistoryResponse(BaseModel):
    profile_id: uuid.UUID
    days: list[ViewingHistoryDay]


# -- Weekly Report schemas --

class TopTitle(BaseModel):
    title_id: uuid.UUID
    title_name: str
    total_minutes: float


class ProfileWeeklyStats(BaseModel):
    profile_id: uuid.UUID
    profile_name: str
    daily_totals: list[dict]
    average_daily_minutes: float
    total_minutes: float
    educational_minutes: float
    limit_usage_percent: float | None
    top_titles: list[TopTitle]


class WeeklyReportResponse(BaseModel):
    week_start: str
    week_end: str
    profiles: list[ProfileWeeklyStats]

File: frontend-client/src/api/parentalControls.ts
import { apiFetch } from './client'

// -- Types --

export interface PinVerifyResponse {
  verified: boolean
  pin_token: string | null
}

export interface PinError {
  detail: string
  // M-09: remaining_attempts removed — server no longer sends attempt count
  locked_until: string | null
}

export interface ViewingTimeConfig {
  profile_id: string
  weekday_limit_minutes: number | null
  weekend_limit_minutes: number | null
  reset_hour: number
  educational_exempt: boolean
  timezone: string
  updated_at: string
}

export interface ViewingTimeConfigUpdate {
  weekday_limit_minutes?: number | null
  weekend_limit_minutes?: number | null
  reset_hour?: number
  educational_exempt?: boolean
  timezone?: string
  pin_token?: string
}

export interface GrantExtraTimeResponse {
  granted_minutes: number | null
  remaining_minutes: number | null
  is_unlimited_override: boolean
}

export interface ViewingHistorySession {
  session_id: string
  title_id: string
  title_name: string
  device_type: string | null
  is_educational: boolean
  started_at: string
  ended_at: string | null
  duration_minutes: number
}

export interface ViewingHistoryDay {
  date: string
  total_minutes: number
  educational_minutes: number
  counted_minutes: number
  sessions: ViewingHistorySession[]
}

export interface ViewingHistoryResponse {
  profile_id: string
  days: ViewingHistoryDay[]
}

export interface TopTitle {
  title_id: string
  title_name: string
  total_minutes: number
}

export interface ProfileWeeklyStats {
  profile_id: string
  profile_name: string
  daily_totals: { date: string; counted_minutes: number; educational_minutes: number }[]
  average_daily_minutes: number
  total_minutes: number
  educational_minutes: number
  limit_usage_percent: number | null
  top_titles: TopTitle[]
}

export interface WeeklyReportResponse {
  week_start: string
  week_end: string
  profiles: ProfileWeeklyStats[]
}

// -- PIN management --

export function createPin(newPin: string, currentPin?: string): Promise<{ detail: string }> {
  return apiFetch('/parental-controls/pin', {
    method: 'POST',
    body: JSON.stringify({ new_pin: newPin, current_pin: currentPin }),
  })
}

export function verifyPin(pin: string): Promise<PinVerifyResponse> {
  return apiFetch('/parental-controls/pin/verify', {
    method: 'POST',
    body: JSON.stringify({ pin }),
  })
}

export function resetPin(password: string, newPin: string): Promise<{ detail: string }> {
  return apiFetch('/parental-controls/pin/reset', {
    method: 'POST',
    body: JSON.stringify({ password, new_pin: newPin }),
  })
}

// -- Viewing Time Config --

export function getViewingTimeConfig(profileId: string): Promise<ViewingTimeConfig> {
  return apiFetch(`/parental-controls/profiles/${profileId}/viewing-time`)
}

export function updateViewingTimeConfig(
  profileId: string,
  config: ViewingTimeConfigUpdate,
): Promise<ViewingTimeConfig> {
  return apiFetch(`/parental-controls/profiles/${profileId}/viewing-time`, {
    method: 'PUT',
    body: JSON.stringify(config),
  })
}

// -- Grant Extra Time --

export function grantExtraTime(
  profileId: string,
  minutes: number | null,
  pin?: string,
  pinToken?: string,
): Promise<GrantExtraTimeResponse> {
  return apiFetch(`/parental-controls/profiles/${profileId}/viewing-time/grant`, {
    method: 'POST',
    body: JSON.stringify({ minutes, pin, pin_token: pinToken }),
  })
}

// -- History & Reports --

export function getViewingHistory(
  profileId: string,
  fromDate?: string,
  toDate?: string,
): Promise<ViewingHistoryResponse> {
  const params = new URLSearchParams()
  if (fromDate) params.set('from_date', fromDate)
  if (toDate) params.set('to_date', toDate)
  const qs = params.toString()
  return apiFetch(`/parental-controls/profiles/${profileId}/history${qs ? `?${qs}` : ''}`)
}

export function getWeeklyReport(): Promise<WeeklyReportResponse> {
  return apiFetch('/parental-controls/weekly-report')
}
